Event based update

gitagogaming opened the pull request:
## Updating MediaMixer to be event-based while reducing performance issues and errors

#### **audioController.py**
1. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/c1f76eac164a1800d257cd6b3d0cacce91a1324d) Added set/increase/decrease option for setDeviceVolume function**
   - Description: Added options to set, increase, or decrease device volume.

2. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/8a20d48e17c07b576b2a280a2388adf52f86f933) Moved setDeviceVolume and setDeviceMute from TPAudioMixer.py**
   - Description: Refactored code to move volume and mute functions to a new file.

3. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/52d579200c610dbb6430dd97ac84979c5f908ba8) Commented out various functions/imports no longer needed; will remove in a later revision once confirmed safe**
   - Description: Cleaned up obsolete functions and imports.

4. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/08577f29fb45bca7d4317ea59b4eff240698fea2) Removed TPClient Import; now passing the object from TPAudioMixer.py**
   - Description: Updated import structure to simplify dependencies.

5. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/ad9c5d7f30d0da8f215918456271e720143dbe61) Added helper function get_session; Changed function name get_process to get_volume to better align with what it actually does**
   - Description: Improved function naming and added session helper function.

#### **audioManager.py**
1. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/40d6a23240b9186eb57aeb0698751350cb090be6) Added some try/except/finally to make sure pythoncom gets uninitialized**
   - Description: Enhanced error handling to ensure proper resource cleanup.

2. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/52d579200c610dbb6430dd97ac84979c5f908ba8) Added a reversed dictionary for output/input devices to use as a reference when needed; Added getDeviceByName function which gets the proper input/output or default device from the audio manager**
   - Description: Implemented new device reference methods for better management.

3. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/03235c678925df3cdbc95b392a7fa1208af5cb6f) Added connector & state update for default input and output devices; Added initialize_default_devices which gets the default devices upon startup; Added fetch_devices which utilizes myAudioUtilities to get the current default device**
   - Description: Improved initialization and updating of default device states.

4. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/ad9c5d7f30d0da8f215918456271e720143dbe61) Initial push; Registers and integrates callback for default device changes as well as any volume/mute changes on any audio device on the PC**
   - Description: Set up callbacks for monitoring and responding to device changes.

#### **TPAudioMixer.py**
1. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/8a20d48e17c07b576b2a280a2388adf52f86f933) Moved VolumeProcess and AppAudioCallBack to AppAudioCallBack.py; Replaced any usage of getActiveExecutablePath to use listener.get_app_path() instead; Moved setDeviceVolume & SetDeviceMute to utils.py**
   - Description: Refactored code for better organization and maintenance.

2. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/52d579200c610dbb6430dd97ac84979c5f908ba8) Fixed handleSettings to remove audio states if user adds to ignore list; Changed removeAudioState to check if it exists before even trying; Changed some logging to debug instead of info; Renamed WinAudioCallback to AppAudioCallback to better represent what it does; Handling new setDeviceMute action; Added SetDeviceVolume func; Simplified/cleaned up some code**
   - Description: Improved error handling and code clarity; updated callback names.

3. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/ef7ab2a763d191a276498b52d53e6021e6a40e17) Created a class for VolumeProcess to reside in as it's only used in a few places within the file; Global no longer needed; Removed getmasterVolume import; Renamed run_callback to run_callbacks since it's doing more than one now**
   - Description: Encapsulated volume process functionality within a class.

4. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/03235c678925df3cdbc95b392a7fa1208af5cb6f) Removed the stateUpdate function as it's no longer needed; this should drastically reduce CPU/memory usage by the user**
   - Description: Eliminated unnecessary function to optimize performance.

#### **appAudioCallBack.py**
1. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/8a20d48e17c07b576b2a280a2388adf52f86f933) New file to separate appAudioCallback and VolumeProcess from TPAudioMixer.py**
   - Description: Created a new file for better modularity.

2. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/40d6a23240b9186eb57aeb0698751350cb090be6) Pep8 formatting for the newer classes and instances of it**
   - Description: Applied PEP8 formatting to improve code readability.

#### **utils.py**
1. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/40d6a23240b9186eb57aeb0698751350cb090be6) Moved dataMapper to utils.py**
   - Description: Centralized data mapping functions in a utility file.

2. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/8a20d48e17c07b576b2a280a2388adf52f86f933) New file; Moved getDevicebyData and getActiveExecutablePath from TPAudioMixer.py**
   - Description: Added new utility file for commonly used functions.

3. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/52d579200c610dbb6430dd97ac84979c5f908ba8) Moved setDeviceVolume & SetDeviceMute to utils.py**
   - Description: Consolidated volume and mute functions into utility file.

#### **windowFocusListener.py**
1. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/03235c678925df3cdbc95b392a7fa1208af5cb6f) New class which uses Windows Event Hooks to listen for window focus event changes and then updates accordingly**
   - Description: Added functionality to track window focus events.

2. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/08577f29fb45bca7d4317ea59b4eff240698fea2) Passed TPClient object and added getLogger for g_log**
   - Description: Integrated logging and client object for event handling.

#### **tppEntry.py**
1. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/c1f76eac164a1800d257cd6b3d0cacce91a1324d) Added hashold functionality to setDeviceVolume device**
   - Description: Implemented hashing functionality for device volume.

2. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/52d579200c610dbb6430dd97ac84979c5f908ba8) Fixed Category Names; Added SetDeviceMute**
   - Description: Corrected category names and added mute functionality.

3. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/ef7ab2a763d191a276498b52d53e6021e6a40e17) Added categories for Default Input/Output devices as well as for the current focused app; Did some basic editing with descriptions; Added Current Focused App Mute State**
   - Description: Expanded categories and updated descriptions.

4. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/1fe465d477dea9c1901d966091ce07b0e503dd83) Renamed updateDevice to updateDeviceChoices to better represent what it does**
   - Description: Renamed function for clarity.

5. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/3c1f3ad00d52e2025879e063ce23bbc36af90ac5) Update tppEntry.py**
   - Description: General updates to the file.

6. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/1daf626692f1ecf6b374b3a026710aa8b09fa84e) Update tppEntry.py**
   - Description: Additional updates to the file.

#### **other**
1. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/35757065b83cd47a2c60a3a38a09e776a81b45cc) Fixed duplicate logger**
   - Description: Addressed issue with duplicate logging.

2. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/dc782585866fbe06c0b3e9d6e86c5b19ce51c715) Removed pygetwindow requirements**
   - Description: Removed unnecessary dependencies.

3. **[LINK](https://github.com/gitagogaming/TouchPortal-Windows-MediaMixer/commit/4d6b9bbbe461b277cfc8fdb61b203a918d) Cleanup, removing print statements**
   - Description: Removed debug print statements for cleaner code.

gitagogaming committed:
Adding AudioManager + Fixes

# audioController.py
- added helper func get_session
- changed func name get_process to get_volume to better align with what it actually does.

# audioManager.py
- initial push
- registers and integrates callback for default device changes as well as any volume/mute changes on any audio device on the PC.

# TPAudioMixer.py
- moved TPClient and g_log to its own file so it can be accessed by audioManager.py

# tppEntry.py
- added master mute state

gitagogaming committed:
updates to audiomanager, TPAudioMixer, tppEntry * added windowFocusListener

# audioManager.py
- added connector & state update for default input and output devices
- added initialize_default_devices which gets the  default devices upon startup.
- added fetch_devices which utilizes myAudioUtiliies to get the current default device

# TPAudioMixer
- Removed the stateUpdate function as its no longer needed. This should drastically reduce the cpu/mem usage by the user, despite it not using much previously.

- added the WindowListener startup inside of runCallback()
- added slider actions for default input/output devices which is bidirectional

# tppEntry.py
- Changed some plugin action and state descriptions(will not affect backwards compatibility.
- added Master Volume Input states

# windowFocusListener
- new class which uses Windows Event Hooks to listen for window focus event changes and then updates accordingly.

gitagogaming committed:
version Update

gitagogaming committed:
Cleanup, removing print statements

gitagogaming committed:
Create build.yml

gitagogaming committed:
Update build.yml

gitagogaming committed:
Update build.yml

gitagogaming committed:
Update build.py

gitagogaming committed:
removed pygetwindow requirements

gitagogaming committed:
fixed duplicate logger

gitagogaming committed:
added try/except removeAudioState()

added try/except when removing audio state.. sometimes it causes an error when app didnt have a volume state

gitagogaming committed:
Update tppEntry.py

gitagogaming committed:
avoiding redundant calls for connector update

added check for last_volume of currentApp.. if changing app windows multiple times that have the same volume, Touchportal would throw an error in notifications about redundant calls

gitagogaming committed:
changing how TPClient is handled, fixed imports

fixed various files where tppEntry was importing *

#audioManager
- removed TPClient Import, now passing the object from TPAudioMixer.py
- removed g_log import, we now just getLogger where needed.

# TPAudioMixer
- instantiate TPClient and pass to other classes as needed
- added shutdown

# deleted TPClient.py

# windowFocusListener
- passed TPClient object and added getLogger for g_log

gitagogaming committed:
new class, deleted old file and added new state & category tppentry

# audiotest.py
- Deleted

# audioManager.py
- uncommented original startup and deleted other leftovers

# TPAudioMixer.py
- created a class for VolumeProcess to reside in as its only used in a few places within the file. Global no longer needed
- removed getmasterVolume import
- renamed run_callback to run_callbacks since its doing more than one now.

# tppentry.py
added categories for Default Input/Output devices as well as for the current focused app.
- Did some basic editing with descriptions
- Added Current Focused App Mute State

gitagogaming committed:
... read the full notes

# audioManager.py
- added a reversed dictionary for output/input devices to use as a reference when needed
- added getDeviceByName function which gets the proper input/output or default device from the audiomanager.

# TPAudioMixer.py
- fixed handleSettings to remove audiostates if user adds to ignore list
- changed removeAudioState to check if it exists before even trying
- changed some logging to debug instead of info
- renamed WinAudioCallback to AppAudioCallback to better represent what it does.
- Handling new setDeviceMute action
- Added SetDeviceVolume func
- simplified/cleaned up some code

# audioController.py
- commented out various functions/imports no longer needed will remove in a later revision once confirm its safe

# tppEntry.py
- Fixed Category Names
- Added SetDeviceMute

gitagogaming committed:
renamed updateDevice & typehinting

# TPAudioMixer.py
- renamed updateDevice to updateDeviceChoices to better represent what it does.

#windowfocusListener
- added typehinting for TPClient

gitagogaming committed:
Update tppEntry.py

gitagogaming committed:
fixed build issue in tppentry.py

due to adding more categories other than main, the tppbuilder failed so we have to manually assign a category for connectors

gitagogaming committed:
cleanup TPAudioMixer.py, new files

# audioController.py
- moved setDeviceVolume and setDevice mute from TPAudioMixer.py

# appAudioCallBack.py
- new file to seperate appAudioCallback and VolumeProcess from TPAudioMixer.py

# utils.py
- new file
- moved getDevicebyData and getActiveExecutablePath from TPAudioMixer.py

# TPAudioMixer.py
- moved VolumeProcess and  AppAudioCallBack to AppAudioCallBack.py
- replaced any usage of getActiveExecutablePath to use listener.get_app_path()instead
- move setDeviceVolume & SetDeviceMute to utils.py

# windowFocusListener.py
- now setting self.current_focused_exe_path for plugins actions to reference with listener.get_app_path()

gitagogaming committed:
hashold for setDeviceVolume &, cleanup

# audioController.py
- added set/increase/decrease option for setDeviceVolume function

# tpaudiomixer.py
- added on-hold for setDeviceVolume
- random cleanup & error prevention

# tppentry.py
- added hasholdfunctionality to setDeviceVolume device

gitagogaming committed:
assuring pythoncom uninitialized

# appAudiocallBack.py
- pep8 formatting for the newer classes and instances of it.

# audioManaager.py
- added some try/except/finally to make sure pythoncom gets uninitialized

# tpaudiomixer
- try except finally for run_callbacks, moving dataMapper to utils.py

# utils.py
- moved dataMapper to utils.py

gitagogaming committed:
readme update

changelog updates and adding #sliders to menu

gitagogaming committed:
readme and image updates

gitagogaming committed:
versioning back to 2.0.0

DamienStaebler commented on src/appAudioCallBack.py:
```diff
@@ -0,0 +1,166 @@
+        self.TPClient = tpClient
+        
+    @classmethod
+    def set_tp_client(cls, client):
```
This is not needed. As you initialize the class, you should pass TouchPortal API Client as an object to your class.

gitagogaming replied:
when you use AppAudioCallBack to create the magicsession object with, it wants an uninitialized class to work with..

when and if i tried to pass args with that and it wouldnt work as the class is initialized.

DamienStaebler commented on src/appAudioCallBack.py:
```diff
@@ -0,0 +1,166 @@
+        
+
+class AppAudioCallBack(MagicSession):
+    TPClient = None
```
Same here, too. I'm not sure exactly the purpose of this.

gitagogaming replied:
issue defining TPClient and others as None or?

DamienStaebler commented on src/appAudioCallBack.py:
```diff
@@ -0,0 +1,166 @@
+    listener = None
+    
+    @classmethod
+    def set_tp_client(cls, client, volume_process, listener):
```
This too

DamienStaebler commented on src/appAudioCallBack.py:
```diff
@@ -0,0 +1,166 @@
+            self.TPClient.stateUpdate(PLUGIN_ID + f".createState.{self.app_name}.volume", str(round(new_volume*100)))
+            #g_log.info(f"{self.app_name} NEW VOLUME", str(round(new_volume*100)))
+            app_connector_id =f"pc_{TP_PLUGIN_INFO['id']}_{TP_PLUGIN_CONNECTORS['APP control']['id']}|{TP_PLUGIN_CONNECTORS['APP control']['data']['appchoice']['id']}={self.app_name}"
+            if app_connector_id in self.TPClient.shortIdTracker :
```
Ok, you can do this instead `if app_connector_shortId := TPClient.shortIdTracker.get(app_connector_id, None):` So you do not have to index multiple times.

DamienStaebler commented on src/appAudioCallBack.py:
```diff
@@ -0,0 +1,166 @@
+            if (activeWindow := self.listener.get_app_path()) != "":
+                current_app_connector_id = f"pc_{TP_PLUGIN_INFO['id']}_{TP_PLUGIN_CONNECTORS['APP control']['id']}|{TP_PLUGIN_CONNECTORS['APP control']['data']['appchoice']['id']}=Current app"
+
+                if current_app_connector_id in self.TPClient.shortIdTracker :
```
Same here as well

DamienStaebler commented on src/audioUtil/audioController.py:
```diff
@@ -83,69 +91,129 @@ def volumeChanger(process, action, value):
+
+
+
+# def volumeChanger(process, action, value):
```
I would just delete these.

gitagogaming replied:
Done

DamienStaebler commented on src/tppEntry.py:
```diff
@@ -54,52 +70,47 @@
+    "Windows Audio": {
+        'category': "main",
+        "id": PLUGIN_ID + ".connector.WinAudio",
+        "name": "Volume Mixer: Windows Volume slider",
```
Maybe change this to "Windows Device Volume Slider"?

gitagogaming replied:
Perhaps, but then App Control should be `App Volume Slider`? 

i dont see much reason to change these as this will not effect the user in any way. Its fairly clear its a slider/connector since when we reference any of them we use TP_PLUGIN_CONNECTORS

DamienStaebler replied:
"App Slider" make sense to me. But when you say "Windows Slider" that could mean anything. But if you add "device" then they know it's targeting specific device's audio

gitagogaming replied:
gotcha.. i just kept it like that since it was the original connector you had commented out.. ill change it up

gitagogaming committed:
creating states for devices other than defaults & efficiency changes

# appAudioCallback.py
- added walrus operator

# audiomanager.py
- creating states & updating connectors for devices when registered in create_callback_for_device
- new function for updating connector

gitagogaming committed:
removed repeated logging

gitagogaming committed:
forgot a walrus! :=

gitagogaming committed:
fixed startup not updating states/sliders properly

fixed startup plugin sync issues for connectors/states

gitagogaming committed:
fixes with current focused app volume/mute

# appaudioCallback.py
- fixed issue where app mute & volume wasnt updated for current focused app if changed

# windowFocusListener
- added self.last_mute and self.current_focused_name to check against before updating state

gitagogaming committed:
fixed performance error when external changes occured

gitagogaming committed:
renaming connectors

Renamed TP_PLUGIN_CONNECTOR["APP control'] to ["APP Volume Slider"]

Renamed TP_PLUGIN_CONNECTOR["Windows Audio"] to ["Device Volume Slider"]

gitagogaming committed:
pluginupdater.py & notifications added

created a small gui interface which when prompted will download the plugin and initiate an install.

gitagogaming committed:
Update audioManager.py

gitagogaming committed:
Revert "pluginupdater.py & notifications added"

This reverts commit af50234b5dd6ae748a7e41532c227123b51bdeca.

gitagogaming committed:
fixed connector and state updates, added set/increase/decrease action choice for setDeviceVolume

#TPAudioMixer.py
- Added volume_action, set/increase/decrease

# audioManager.py
- The default output connector is now correctly updated when the plugin starts.
- Default devices are properly initialized, ensuring that mute and volume settings are updated consistently.
- Now using stateupdateMany to avoid sending too many updates at once when initializing default devices
